monitor/common.go

```diff
@@ -54,63 +55,67 @@ func (m *Monitor) isPusher() bool {
 	return pusher.ValidatorAddr.Equals(m.operator.Key.GetAddress())
 }
 
-func (m *Monitor) isPullerOrOwner(candidate mainchain.Addr) bool {
-	return m.isPuller() || candidate == m.ethClient.Address
-}
-
 // Is the current node the guard to submit state proof
-func (m *Monitor) isRequestGuard(request *guard.Request, eventBlockNumber uint64) bool {
-	requestGuards := request.RequestGuards
-	if len(requestGuards) == 0 {
-		log.Debug("no request guards")
+func (m *Monitor) isCurrentGuard(request *guard.Request, eventBlockNumber uint64) bool {
+	assignedGuards := request.AssignedGuards
+	if len(assignedGuards) == 0 {
+		log.Debug("no assigned guards")
 		return false
 	}
 
-	blkNum := m.ethMonitor.GetCurrentBlockNumber().Uint64()
+	blkNum := m.getCurrentBlockNumber().Uint64()
 	blockNumberDiff := blkNum - eventBlockNumber
-	guardIndex := uint64(len(requestGuards)+1) * blockNumberDiff / request.DisputeTimeout
+	guardIndex := uint64(len(assignedGuards)+1) * blockNumberDiff / request.DisputeTimeout
 
-	log.Infoln("IsRequestGuard", blkNum, eventBlockNumber, guardIndex, requestGuards)
 	// All other validators need to guard
-	if guardIndex >= uint64(len(requestGuards)) {
+	if guardIndex >= uint64(len(assignedGuards)) {
+		log.Debugln("Assigned guard:", blkNum, eventBlockNumber, guardIndex)
 		return true
 	}
+	log.Debugln("Assigned guard:", blkNum, eventBlockNumber, guardIndex, assignedGuards[guardIndex].String())
 
-	return requestGuards[guardIndex].Equals(m.operator.Key.GetAddress())
+	return assignedGuards[guardIndex].Equals(m.operator.Key.GetAddress())
 }
 
-func (m *Monitor) getRequest(channelId []byte, simplexReceiver string) (guard.Request, error) {
-	return guard.CLIQueryRequest(m.operator.CliCtx, guard.RouterKey, channelId, simplexReceiver)
+func (m *Monitor) getGuardRequest(channelId []byte, simplexReceiver string) (*guard.Request, error) {
+	request, err := guard.CLIQueryRequest(m.operator.CliCtx, guard.RouterKey, channelId, simplexReceiver)
+	if err != nil {
+		return nil, err
+	}
+	return &request, nil
```

Comment:
Is this change just for performance optimization?

Reply:
Yes. In Go (and many other) programs, we should return pointer of struct in most cases to avoid deep copy. Though in most blockchain systems memory copy won't be the performance bottleneck, it is still good to comply with the correct programming model